keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,570 @@
+This could be mitigated by using iptables to implement the AllPorts logic.
+
+* A known issue with [host services being accessible via ClusterIP in ipvs mode](https://github.com/kubernetes/kubernetes/issues/72236) could be mitigated by AllPorts support.
+If a ClusterIP service is created using `allPorts` set to true and sshd on the host listens on 0.0.00:22, traffic to `<ClusterIP>:22` will only go to backend pods.
+If the service were not using `allPorts`, and did not specify port 22 in the Port list, sshd would be exposed by connecting to `<ClusterIP>:22`.
+This is because the clusterIP is assigned to an ipvs interface in the host namespace.
+
+## Design Details
+
+Changes are required to APIServer validation, kube-proxy and controllers that use the ServicePort field.
+
+1) New Validation checks:
+
+   * `allPorts` can be set to True only for ClusterIP(non-headless) and LoadBalancer services.
+   * The `ports` array should be empty when `allPorts` is set to True.
+
+2) Kube-Proxy should configure iptables/ipvs rules by skipping port/protocol filter, if `allPorts` is true.
+
+3) LoadBalancer controllers should create LoadBalancer resources with the appropriate port values.
+
+4) Endpoints and EndpointSlices controller should create Endpoints with empty port values.
+
+5) DNS Providers should handle empty ports in Services and Endpoints.
+   There will be no DNS SRV Records for Services with `allPorts` set.
+   coreDNS handles empty ports in [services](https://github.com/coredns/coredns/blob/09b63df9c1584bb5389d1b681698631bcd7c19e1/plugin/kubernetes/kubernetes.go#L577) and [endpoints.](https://github.com/coredns/coredns/blob/09b63df9c1584bb5389d1b681698631bcd7c19e1/plugin/kubernetes/kubernetes.go#L559)
+   kube-dns also handles empty ports in [services](https://github.com/kubernetes/dns/blob/077a43e83e648ba5f04bae18ffcb824edc9db967/pkg/dns/dns.go#L506) and [endpoints.](https://github.com/kubernetes/dns/blob/077a43e83e648ba5f04bae18ffcb824edc9db967/pkg/dns/dns.go#L541)
+   There is a warning in kube-dns for [empty port services](https://github.com/kubernetes/dns/blob/077a43e83e648ba5f04bae18ffcb824edc9db967/pkg/dns/dns.go#L320) that could be removed after GA of AllPorts.
+
+6) There will be no environment variable of the form "_SERVICE_PORT" for services with `allPorts` set. This [codepath](https://github.com/kubernetes/kubernetes/blob/e1f971d5c2a1002c4e90471d064f87f297740aba/pkg/kubelet/envvars/envvars.go#L48) currently assumes non-zero ports and that will be updated.
+
+#### Open Questions to be resolved:
+
+1) How should IPVS implementation be handled?
+Options are to a) use iptables for DNAT b) use ipvs rules that results in same 5-tuple requests being sent to the same backend pod. c) create allPorts services as "fwmark-service" and assign a unique mark for each of them.
+
+2) Identfy how Service Mesh(Istio)/Calico/MetalLB can support AllPorts.
```

Comment:
Can I have a UDP all-ports Service and a TCP all-ports Service on the same ClusterIP?